### DISCUSSION TITLE
Meshing Quandary: 1/3 2/3 Rule and SmoothMeshLines() Function Clash

### DISCUSSION THREAD

**Discussion by Sarajevo67** [thliebig/openEMS-Project#102] [Q&A Usage] [answered]
Hello everyone,

I've been working on finding a solution for a 3-segment SL/MSL (Stepped Impedance Matching Transformer) as described in my earlier post. While parsing SL/MSL Tutorials and experimenting with various setups, I came across the 1/3 2/3 ratio meshing rule and confusing behavior when using the SmoothMeshLines() function.

Here's an example of what I encountered:

```
resolution = 2400;
MSL_width = 512;

>> [0 MSL_width/2+[-resolution/3 +resolution/3*2]/4]
ans =
     0    56   656
```

As you can see, the vector for the 1/3 2/3 rule mesh is correctly defined as [0 56 656].

However, when I used the SmoothMeshLines() function, it changed the mesh points in the same line, resulting in a "freestyle" mesh:

```
>> mesh.y = SmoothMeshLines( [0 MSL_width/2+[-resolution/3 +resolution/3*2]/4], resolution/4 , 1.5 ,0)
mesh =
  scalar structure containing the fields:
    y =
              0    56.0000   140.0000   266.0000   455.0000   555.5000   656.0000

```

The SmoothMeshLines() function seems to have altered ratio of the mesh lines positions, which is not what I expected. Consequently, the distance between (256 and 140) is not half of the distance between (266 and 256).

Could someone kindly help me understand if I missed something or if there's a better way to achieve desired meshing?

Thank you for your assistance and insights!

**Comment by VolkerMuehlhaus** [comment]
You seem to confuse the **local** mesh resolution (used in thirds rule) and the **maximum** mesh size.

Parameter "resolution" has a misleading name, that is really the maximum possible mesh size in areas where we don't need to resolve geometry details. This is calculated as a fraction of the effective wavelength in your dielectrics, e.g. lambda/20.

For transmission lines, we want to have ~ 5 cells across the line width, this determines the local mesh size for line width. The cells at the outer boundary are offset using the thirds rule, using this local mesh size.

**Reply by Sarajevo67** [reply]
Thanks for your _5 cells across the line width_ clarification. It helps me a lot.

For this question purpose, I still have some doubts, because I used excerpt from Tutorial: MSL_NotchFilter and I'm focused just on the points around line edge.

**Just look, what happened around line edge (MSL_width/2=256) in my example: "inside" mesh line is at 140.0000 and outside mesh line is at 266.0000** 

Is it expected behavior?
**This mesh construct is used in almost all SL/MSL tutorials I found:**
```
%% setup CSXCAD geometry & mesh %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
CSX = InitCSX();
resolution = c0/(f_max*sqrt(substrate_epr))/unit /50; % resolution of lambda/50
mesh.x = SmoothMeshLines( [0 MSL_width/2+[2*resolution/3 -resolution/3]/4], resolution/4, 1.5 ,0 );
mesh.x = SmoothMeshLines( [-MSL_length -mesh.x mesh.x MSL_length], resolution, 1.5 ,0 );
mesh.y = SmoothMeshLines( [0 MSL_width/2+[-resolution/3 +resolution/3*2]/4], resolution/4 , 1.5 ,0);
mesh.y = SmoothMeshLines( [-15*MSL_width -mesh.y mesh.y stub_length+[-resolution/3 +resolution/3*2]/4 15*MSL_width+stub_length], resolution, 1.3 ,0);
mesh.z = SmoothMeshLines( [linspace(0,substrate_thickness,5) 10*substrate_thickness], resolution );
CSX = DefineRectGrid( CSX, unit, mesh );
```
Where is initial 'mesh.y' vector, **constructed for thirds rule purpose** (I guess), IS later incorporated in final 'mesh.y', therefore it uses: 

`[0 MSL_width/2+[-resolution/3 +resolution/3*2]/4]`

as SmoothMeshLines() function argument:

`mesh.y = SmoothMeshLines( [0 MSL_width/2+[-resolution/3 +resolution/3*2]/4], resolution/4 , 1.5 ,0);`

**Which, apparently, broke thirds rule, as described in initial post above.**

Second  SmoothMeshLines() call, symetricaly incorporates "broken" (-mesh.y mesh.y) in the final 'mesh.y' vector:

`mesh.y = SmoothMeshLines( [-15*MSL_width -mesh.y mesh.y stub_length+[-resolution/3 +resolution/3*2]/4 15*MSL_width+stub_length], `

I hope I'm not hallucinating :-)

**Reply by VolkerMuehlhaus** [reply, marked as answer]
Maybe you misunderstood me - the numbers in your example make no sense because you create the third rules from the wrong "resolution" value.  It must be one third of the **local** mesh density used to mesh the line width. 

In your example, your 1/3 offset is larger than the line width because you calculate it from lambda/50. That goes wrong because the 1/3 mesh offset is larger than the entire line width...

In your example with 512µm line width, mesh size across line width should be around 100µm, and 1/3 rule refers to this local mesh size. It does not refer to the much larger **maximum** mesh size (misleading name "resolution" in your code) calculated from lambda/50!

**Reply by Sarajevo67** [reply]
I understand you, but original resolution=2400; value is not from my head, but **from Tutorial**, calculated in:

`resolution = c0/(f_max*sqrt(substrate_epr))/unit /50;`

Which, for f_max=1.30585e9 and substrate_epr=3.66; gives resolution=2400. Later it is divided by /3, /4 and /3/4 in the code.
My mistake, obviously, is that I decreased f_max, expecting that code will adjust rest of design.
None of this is from my head. I am still learning and that's why I was confused.

**Thanks on your time, I realized that I have to rewrite the code, when changing frequency range.**

**Reply by VolkerMuehlhaus** [reply]
Yes, the issue with this calculation is that "resolution" is calculated from frequency, and ignores dimensions of your drawn structure. 

This name "resolution" can easily be misunderstood, that is why I would recommend to differentiate between geometry-based local mesh resolution and maximum mesh size required in "empty" areas.

**Reply by Sarajevo67** [reply]
Thanks, I see now, that I wrongfully expected that meshing in and around lines is automated in tutorials. I probably will write separate function for automated meshing around lines, which I need, anyway, for meshing problem in my 3 segment SL/MSL transformer problem.

As soon as I find a way to "present" antenna impedance readings to the transformer... :-)
